### PR TITLE
Add forward AD inplace check and fix codegen

### PR DESCRIPTION
**Pull request by albanD**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #59993 All remaining linear/element-wise formulas
* #60499 Fix out= variant forward grad detection
* **#60498 Add forward AD inplace check and fix codegen**

Differential Revision: [D29914593](https://our.internmc.facebook.com/intern/diff/D29914593)

### CODE REVIEW

**Comment by albanD** [comment]
You can see codegened code [here](https://github.com/albanD/pytorch/commits/codegen_diff) which is the same commits but with the generated files checked in.

**Comment by zou3519** [comment]
PR body should contain more description about how the forward AD inplace check works and what fixes were applied. I think I know the gist from our discussions but a summary is nice and anyone who comes along in the future who finds this commit may be confused

File: tools/autograd/gen_variable_type.py [hunk 1]
```diff
@@ -583,6 +588,23 @@ def emit_check_no_requires_grad(
             body.append(f'check_no_requires_grad({arg_name}, "{arg_name}", "{name}");')
         return body
 
+    def emit_orignal_self_definition() -> List[str]:
+        body: List[str] = []
+        if inplace:
+            body.append('c10::optional<at::Tensor> original_self;')
```
Comment (zou3519) [inline tools/autograd/gen_variable_type.py:594]: I remember Ed saying something a while ago about splitting off forward-mode AD into its own DispatchKey (to be separate from reverse-mode AD). Some of the optimizations we're doing here (clone only once if both forward mode and reverse mode need a close of the input) are difficult to do if we do split forward-mode AD into different dispatch keys. Is this a concern?

File: tools/autograd/gen_variable_type.py [hunk 2]
```diff
@@ -583,6 +588,23 @@ def emit_check_no_requires_grad(
             body.append(f'check_no_requires_grad({arg_name}, "{arg_name}", "{name}");')
         return body
 
+    def emit_orignal_self_definition() -> List[str]:
```
Comment (zou3519) [inline tools/autograd/gen_variable_type.py:591]: nit: original

File: tools/autograd/gen_variable_type.py (hunk 1 above)
Comment (albanD) [reply tools/autograd/gen_variable_type.py:594]: We might have to give up such optimizations when we move to a different dispatch key indeed.
We might be able to unpack the SavedVariable in the Node we just created, but that would be more complex than the current code for sure.

File: tools/autograd/gen_variable_type.py [hunk 3]
```diff
@@ -304,6 +304,11 @@
 }
 """)
 
+FW_DERIVATIVE_BOOLEAN_TEMPLATE = CodeTemplate("""\
+auto ${boolean_var_name} = ${cond};
+(void)${boolean_var_name};
+""")
```
Comment (zou3519) [inline tools/autograd/gen_variable_type.py:310]: Is there a way to check that the compiler optimized away this code if it's not used?
```
auto _any_has_forward_grad_result = isFwGradDefined(self) || isFwGradDefined(other);
(void)_any_has_forward_grad_result;
```
I remember sometimes the compiler gets confused about if it's safe to optimize away things (e.g. if the compiler thinks isFwGradDefined has any side effects)

File: tools/autograd/gen_variable_type.py [hunk 4]
```diff
@@ -583,6 +588,23 @@ def emit_check_no_requires_grad(
...
+                if derivative.required_original_self_value:
+                    all_forward_grad_cond.append(get_any_has_forward_grad_name(derivative.var_name))
+
+            if all_forward_grad_cond:
+                body.append(f'if ({" || ".join(all_forward_grad_cond)}) {{')
+                body.append('  original_self = self.clone();')
+                body.append('}')
+
+        return body
+
```
Comment (zou3519) [inline tools/autograd/gen_variable_type.py:607]: We probably want a big note somewhere about what the structure of a (for example) VariableType::mul_ code looks like. Some things I had to remind myself about include: we need to clone `self` before the redispatch because otherwise self can get mutated

File: tools/autograd/gen_variable_type.py [hunk 5]
```diff
@@ -583,6 +588,23 @@ def emit_check_no_requires_grad(
...
+            body.append('c10::optional<at::Tensor> original_self;')
+
+            all_forward_grad_cond = []
+            for derivative in fw_derivatives:
+                if derivative.required_original_self_value:
+                    all_forward_grad_cond.append(get_any_has_forward_grad_name(derivative.var_name))
+
+            if all_forward_grad_cond:
+                body.append(f'if ({" || ".join(all_forward_grad_cond)}) {{')
+                body.append('  original_self = self.clone();')
```
Comment (zou3519) [inline tools/autograd/gen_variable_type.py:603]: Is... the first argument always called self?

File: tools/autograd/gen_variable_type.py [hunk 6]
```diff
@@ -618,9 +642,13 @@ def save_variables(
                 expr = f'{expr}.has_value() ? c10::optional<std::string>(std::string({expr}.value())) : c10::nullopt'
             guard = guard_for(arg)
             if guard is None:
+                if stmts_prepend:
+                    stmts.append(f'{stmts_prepend};')
                 stmts.append(f'grad_fn->{name} = {expr};')
             else:
                 stmts.append(f'if ({guard}) {{')
+                if stmts_prepend:
+                    stmts.append(f'  {stmts_prepend};')
```
Comment (zou3519) [inline tools/autograd/gen_variable_type.py:651]: I'm not familiar with this code -- what is an example of the `guard` here?

File: tools/autograd/gen_variable_type.py [hunk 7]
```diff
@@ -759,12 +780,40 @@ def emit_fw_derivatives() -> List[str]:
                                        'formula has been defined for it. This case should only happen for function that '
                                        'take a single TensorList as input. All other cases are not supported right now.')
                 requires_fw_grad = "true"
+
+            content.append(FW_DERIVATIVE_BOOLEAN_TEMPLATE.substitute(
```
Comment (zou3519) [inline tools/autograd/gen_variable_type.py:784]: Nit: Could we inline FW_DERIVATIVE_BOOLEAN_TEMPLATE here? the snippet is pretty small and inlining would improve readability

File: tools/codegen/api/autograd.py [hunk 8]
```diff
@@ -67,6 +67,14 @@ class ForwardDerivative:
     # Inputs for which the primal is required for this formula
     required_inputs_primal: Optional[Tuple[str, ...]]
 
+    # Flag to specify if this formula requires the original value of self
+    # This is only used by inplace operations
+    required_original_self_value: bool
+
+    # If this formula is specified by the user or if we are re-using the
```
Comment (zou3519) [inline tools/codegen/api/autograd.py:74]: nit: "specified by the user" -> "specified in derivatives.yaml"? A user of PyTorch cannot specify a formula, right?

File: tools/codegen/api/autograd.py [hunk 9]
```diff
@@ -67,6 +67,14 @@ class ForwardDerivative:
...
+    # out of place formula for inplace
+    is_exact_match: bool
```
Comment (zou3519) [inline tools/codegen/api/autograd.py:76]: Would prefer a better name on this. How about "reuses_outplace_formula"?

File: tools/codegen/api/autograd.py [hunk 10]
```diff
@@ -259,27 +267,53 @@ def find_info(f: NativeFunction) -> Tuple[Optional[DifferentiabilityInfo], bool]
...
-
-                if re.search(IDENT_REGEX.format("self"), fw_info.formula):
-                    raise RuntimeError(f'The formula for "{f.func.name}" is using the original value of self that is being '
-                                       'modified inplace. This would lead to wrong forward gradients. Please use "result" in '
-                                       'the formula only.')
-
-                # replace "result" from the formula by self
+                formula = fw_info.formula
+
+                def update_for_original(formula: str, postfix: str) -> str:
```
Comment (zou3519) [inline tools/codegen/api/autograd.py:272]: nit: I had to read the use sites to figure out what this is doing. How about calling this "replace_self_with_original_self"?

File: tools/codegen/api/autograd.py [hunk 11]
```diff
@@ -259,27 +267,53 @@ def find_info(f: NativeFunction) -> Tuple[Optional[DifferentiabilityInfo], bool]
...
+                        # When the original formula is out of place, we save a clone of the primal
+                        # value to be able to access this value if needed
+                        # replace "self_p"/"self_t" from the formula by "original_self_p"/"original_self_t"
+                        formula = update_for_original(formula, "_p")
+                        formula = update_for_original(formula, "_t")
+
+                if re.search(IDENT_REGEX.format("original_self_p"), formula):
+                    required_original_self_value = True
+                else:
+                    required_original_self_value = False
```
Comment (zou3519) [inline tools/codegen/api/autograd.py:293]: nit: `required_original_self_value = re.search(IDENT_REGEX.format("original_self_p"), formula)`

File: tools/codegen/api/autograd.py [hunk 12]
```diff
@@ -259,27 +267,53 @@ def find_info(f: NativeFunction) -> Tuple[Optional[DifferentiabilityInfo], bool]
...
+
+                # replace "result" from the formula by "self_p"
                 def repl(m: Match[str]) -> str:
-                    return f'{m.group(1)}self{m.group(2)}'
-                formula = re.sub(IDENT_REGEX.format("result"), repl, fw_info.formula)
+                    return f'{m.group(1)}self_p{m.group(2)}'
+                formula = re.sub(IDENT_REGEX.format("result"), repl, formula)
```
Comment (zou3519) [inline tools/codegen/api/autograd.py:298]: nit: repl -> replace_result_with_self_p

File: tools/codegen/api/autograd.py [hunk 13]
```diff
@@ -259,27 +267,53 @@ def find_info(f: NativeFunction) -> Tuple[Optional[DifferentiabilityInfo], bool]
...
+
+                required_primals = fw_info.required_inputs_primal
+                if re.search(IDENT_REGEX.format("self_p"), formula):
```
Comment (zou3519) [inline tools/codegen/api/autograd.py:301]: I'm confused: didn't we replace all instances of `self_p` with `original_self_p`?

File: tools/autograd/gen_variable_type.py (hunk 5 above)
Comment (zou3519) [reply tools/autograd/gen_variable_type.py:603]: It looks like it always is :O

Comment (albanD) [reply tools/autograd/gen_variable_type.py:603]: This is for inplace ops. And for these, it always is. And it is always the argument that is modified inplace.

File: tools/autograd/gen_variable_type.py (hunk 6 above)
Comment (albanD) [reply tools/autograd/gen_variable_type.py:651]: These guards are used when there are multiple formulas for a given function so that we can avoid saving Tensors when that input doesn't require gradients.
For example, for addmm:
```c++
    if (grad_fn->should_compute_output(1)) {
      grad_fn->mat2_ = SavedVariable(mat2, false);
    }
    grad_fn->alpha = alpha;
    if (grad_fn->should_compute_output(2)) {
      grad_fn->mat1_ = SavedVariable(mat1, false);
    }
```

File: tools/codegen/api/autograd.py (hunk 9 above)
Comment (albanD) [reply tools/codegen/api/autograd.py:76]: Yes.
I re-used the one that was already in that function. But in this context it can be improved indeed.

File: tools/codegen/api/autograd.py (hunk 13 above)
Comment (albanD) [reply tools/codegen/api/autograd.py:301]: But then we changed all the "result" to "self_p".
Let me add a comment at the beginning explaining the overall logic for this function

File: tools/autograd/gen_variable_type.py (hunk 3 above)
Comment (albanD) [reply tools/autograd/gen_variable_type.py:310]: From checking the assembly code, the calls to isFwGradDefined are not optimized out indeed.
I'll add the appropriate logic to avoid these calls for out= functions

Comment (zou3519) [reply tools/autograd/gen_variable_type.py:310]: RIP, I wish the compiler were smarter

File: tools/codegen/api/autograd.py (hunk 12 above)
Comment (albanD) [reply tools/codegen/api/autograd.py:298]: Not sure, these small functions are just for the matching. I added a comment above explaining what the pair regex/function is doing. Is that good?

File: tools/autograd/gen_variable_type.py (hunk 3 above)
Comment (albanD) [reply tools/autograd/gen_variable_type.py:310]: I think it could be if they were inlined and only calling simple functions, but that's not the case here :/

File: tools/autograd/derivatives.yaml [hunk 14]
```diff
@@ -761,7 +762,7 @@
 
 - name: polygamma_(Tensor(a!) self, int n) -> Tensor(a!)
   self: grad * polygamma(n + 1, self)
-  result: auto_element_wise
+  result: self_t.mul_(polygamma(n + 1, original_self_p))
```
Comment (soulitzer) [inline tools/autograd/derivatives.yaml:765]: There used to be some `conj` here? But I'm guessing that's okay because polygamma doesn't support complex
It also looks like we avoid doing mul by self_t out of place and then copy_
Are these optimizations that we could automate with auto_element_wise/auto_linear as well?

**Review by soulitzer — APPROVED** [review]
Looks good to me! Just have some small questions.

File: tools/autograd/gen_variable_type.py [hunk 15]
```diff
@@ -614,6 +614,23 @@ def emit_check_no_requires_grad(
             body.append(f'check_no_requires_grad({arg_name}, "{arg_name}", "{name}");')
         return body
 
+    def emit_original_self_definition() -> List[str]:
```
[the remark this answers is not in the document]
Comment (albanD) [reply tools/autograd/gen_variable_type.py:617]: The clone is saved as a SavedVariable on the grad_fn. So it outlives that function.
But we do keep one extra refcount to that Tensor longer indeed. I don't think it matters here though as nothing will check the SavedVariable before a hook is used or the backward happens.

File: tools/codegen/api/autograd.py [hunk 16]
```diff
@@ -257,29 +265,71 @@ def find_info(f: NativeFunction) -> Tuple[Optional[DifferentiabilityInfo], bool]
             forward_derivatives = info.forward_derivatives
 
             if f.func.kind() == SchemaKind.inplace:
+                # For inplace functions there is a little bit of work to do:
```
[the remark this answers is not in the document]
Comment (albanD) [reply tools/codegen/api/autograd.py:268]: The post_process function is called for every formula in the derivatives.yaml file. So at that time, we don't know yet what is the function we're currently writing the code for.
This function here is the first one where we know both the exact function we're writing code for (and if it is inplace) and which formula we're using for it (and if it is an exact match or not).

File: tools/autograd/derivatives.yaml (hunk 14 above)
Comment (albanD) [reply tools/autograd/derivatives.yaml:765]: We have a grand total of 10 functions that have formulas for the inplace version specifically. So I don't think we want to bother adding such a logic to these automation.

File: tools/autograd/gen_variable_type.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (soulitzer) [reply tools/autograd/gen_variable_type.py:617]: Makes sense, I didn't realize that this would be considered `output=False` and lead to the `save_original=True` logic in SavedVariable

File: tools/autograd/derivatives.yaml (hunk 14 above)
Comment (soulitzer) [reply tools/autograd/derivatives.yaml:765]: Ah, good point
What was is the reasoning behind having functions have formulas specifically for inplace anyway?

Comment (albanD) [reply tools/autograd/derivatives.yaml:765]: A few of them are because there is no out-of-place version.
A few of them is for perf to avoid input cloning by providing an alternative implementation that only makes use for the output.

**Comment by albanD** [comment]
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29914593).

**Comment by albanD** [comment]
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29914593).

File: tools/codegen/api/autograd.py [hunk 17]
```diff
@@ -257,29 +265,71 @@ def find_info(f: NativeFunction) -> Tuple[Optional[DifferentiabilityInfo], bool]
...
+                #      populated by cloned version of the original input (either the clone done by the backward AD
+                #      logic if self is also used in a backward formula or a special clone that we add).
+                #  2) At this point, there cannot be a self_p in the formula.
+                #  3) Change "result" into "self_p" as by design, in the inplace function codegen, the result is
+                #     simply called self (as it is modified inplace).
+                #  4) Update the required primals data in case it used to contain "result" but should now contain
+                #     "self"
+                #  5) If it is not an exact match, the user formula is not modifying the existing forward grad
+                #     inplace as it should. So add some code that makes sure that we do so if the forward grad
+                #     already exists.
```
Comment (zou3519) [inline tools/codegen/api/autograd.py:285]: nit: an example here would help, it's hard to read through a chunk of text without being able to visualize what happens

**Review by zou3519 — COMMENTED** [review]
Sorry for the delayed review -- only had a comment on the comment but otherwise this lgtm as well
